File: packages/shared-types/src/inventory.ts
import { AuditModel } from './audit';

export type ItemCategory = 'PROTEIN' | 'CARBOHYDRATE' | 'VEGETABLE' | 'FRUIT' | 'DAIRY' | 'SPICE' | 'PACKAGING' | 'OTHER';

export interface Item extends AuditModel {
  sku: string;
  name: string;
  category: ItemCategory;
  unit: string; // e.g. "kg", "gram", "pcs", "liter"
  min_stock_threshold: number;
  is_perishable: boolean;
  total_stock?: number;
  batches?: ItemBatch[];
}

export interface ItemBatch extends AuditModel {
  item_id: string;
  batch_code: string;
  expiry_date: string; // YYYY-MM-DD
  unit_cost: number; // in IDR (Decimal/Float)
  initial_qty: number;
  current_qty: number;
  received_date: string;
  supplier_name?: string;
  item?: Item;
}

export type MovementType = 'IN' | 'OUT' | 'ADJUSTMENT' | 'WASTE';
export type ReferenceType = 'PURCHASE_RECEIPT' | 'MEAL_PRODUCTION' | 'DISTRIBUTION' | 'STOCK_OPNAME' | 'EXPIRED_DISPOSAL';

export interface StockMovement extends AuditModel {
  item_batch_id: string;
  movement_type: MovementType;
  quantity: number;
  reference_type: ReferenceType;
  reference_id: string;
  notes?: string;
  unit_cost_snapshot: number;
  total_cost_snapshot: number;
  item_batch?: ItemBatch;
}

export interface CreateItemRequest {
  sku: string;
  name: string;
  category: ItemCategory;
  unit: string;
  min_stock_threshold?: number;
  is_perishable: boolean;
}

export interface CreateBatchRequest {
  item_id: string;
  batch_code?: string;
  expiry_date: string;
  unit_cost: number;
  quantity: number;
  supplier_name?: string;
}

export interface StockInBatchInput {
  item_id: string;
  batch_code?: string;
  expiry_date: string;
  unit_cost: number;
  quantity: number;
  supplier_name?: string;
}

export interface StockOutRequest {
  item_id: string;
  requested_qty: number;
  reference_type: ReferenceType;
  reference_id: string;
  notes?: string;
}

export interface DepletedBatchAllocation {
  batch_id: string;
  batch_code: string;
  expiry_date: string;
  depleted_qty: number;
  remaining_qty: number;
  unit_cost: number;
  subtotal_cost: number;
}

export interface StockOutResult {
  item_id: string;
  item_name: string;
  total_qty_depleted: number;
  total_cost: number;
  allocations: DepletedBatchAllocation[];
}

export interface ItemStockSummary {
  id: string;
  sku: string;
  name: string;
  category: ItemCategory;
  unit: string;
  min_stock_threshold: number;
  is_perishable: boolean;
  total_stock: number;
  active_batch_count: number;
  earliest_expiry?: string | null;
  is_low_stock: boolean;
  batches?: ItemBatch[];
}

File: apps/backend/internal/models/inventory.go
package models

import (
	"time"

	"github.com/google/uuid"
)

type ItemCategory string

const (
	CategoryProtein      ItemCategory = "PROTEIN"
	CategoryCarbohydrate ItemCategory = "CARBOHYDRATE"
	CategoryVegetable    ItemCategory = "VEGETABLE"
	CategoryFruit        ItemCategory = "FRUIT"
	CategoryDairy        ItemCategory = "DAIRY"
	CategorySpice        ItemCategory = "SPICE"
	CategoryPackaging    ItemCategory = "PACKAGING"
	CategoryBeverage     ItemCategory = "BEVERAGE"
	CategoryOther        ItemCategory = "OTHER"
)

type MovementType string

const (
	MovementIn         MovementType = "IN"
	MovementOut        MovementType = "OUT"
	MovementAdjustment MovementType = "ADJUSTMENT"
	MovementWaste      MovementType = "WASTE"
)

type ReferenceType string

const (
	RefPurchaseReceipt      ReferenceType = "PURCHASE_RECEIPT"
	RefMealProduction       ReferenceType = "MEAL_PRODUCTION"
	RefDistribution         ReferenceType = "DISTRIBUTION"
	RefStockOpname          ReferenceType = "STOCK_OPNAME"
	RefExpiredDisposal      ReferenceType = "EXPIRED_DISPOSAL"
	RefDailyReconciliation ReferenceType = "DAILY_RECONCILIATION"
)

type Item struct {
	AuditModel
	SKU               string       `gorm:"type:varchar(64);uniqueIndex;not null" json:"sku"`
	Name              string       `gorm:"type:varchar(255);not null" json:"name"`
	Category          ItemCategory `gorm:"type:varchar(32);not null" json:"category"`
	Unit              string       `gorm:"type:varchar(32);not null" json:"unit"` // e.g. "kg", "gram", "pcs"
	GramsPerUnit      float64      `gorm:"type:numeric(15,4);not null;default:1" json:"grams_per_unit"` // Used for converting recipe portions to inventory units
	MinStockThreshold float64      `gorm:"type:numeric(15,4);not null;default:0" json:"min_stock_threshold"`
	IsPerishable      bool         `gorm:"type:boolean;not null;default:true" json:"is_perishable"`
	Batches           []ItemBatch  `gorm:"foreignKey:ItemID" json:"batches,omitempty"`
}

// ItemBatch represents a physical lot of inventory for FEFO (First Expired First Out) stock management
type ItemBatch struct {
	AuditModel
	ItemID       uuid.UUID `gorm:"type:uuid;not null;index:idx_item_batches_fefo,priority:1" json:"item_id"`
	Item         *Item     `gorm:"foreignKey:ItemID" json:"item,omitempty"`
	BatchCode    string    `gorm:"type:varchar(64);uniqueIndex;not null" json:"batch_code"`
	ExpiryDate   time.Time `gorm:"type:date;not null;index:idx_item_batches_fefo,priority:2;index" json:"expiry_date"`
	UnitCost     float64   `gorm:"type:numeric(15,4);not null" json:"unit_cost"`
	InitialQty   float64   `gorm:"type:numeric(15,4);not null" json:"initial_qty"`
	CurrentQty   float64   `gorm:"type:numeric(15,4);not null;index" json:"current_qty"`
	ReceivedDate time.Time `gorm:"type:date;not null" json:"received_date"`
	SupplierName string    `gorm:"type:varchar(255)" json:"supplier_name,omitempty"`
}

// StockMovement tracks exact batch depleting or replenishing for audit trails
type StockMovement struct {
	AuditModel
	ItemBatchID       uuid.UUID     `gorm:"type:uuid;not null;index" json:"item_batch_id"`
	ItemBatch         *ItemBatch    `gorm:"foreignKey:ItemBatchID" json:"item_batch,omitempty"`
	MovementType      MovementType  `gorm:"type:varchar(32);not null" json:"movement_type"`
	Quantity          float64       `gorm:"type:numeric(15,4);not null" json:"quantity"`
	ReferenceType     ReferenceType `gorm:"type:varchar(64);not null;index" json:"reference_type"`
	ReferenceID       string        `gorm:"type:varchar(128);not null;index" json:"reference_id"`
	Notes             string        `gorm:"type:text" json:"notes,omitempty"`
	UnitCostSnapshot  float64       `gorm:"type:numeric(15,4);not null" json:"unit_cost_snapshot"`
	TotalCostSnapshot float64       `gorm:"type:numeric(15,4);not null" json:"total_cost_snapshot"`
}
